### ISSUE TITLE
lm_lin

### ISSUE THREAD

**Issue by graemeblair** [DeclareDesign/estimatr#13]
A new function which implements the estimator proposed in [Lin (2013)](http://www.stat.berkeley.edu/~winston/agnostic.pdf). It is basically a thin wrapper that pre-processes the formula to do the transformation suggested in the paper and then sends to `lm_robust`. The options should be:

`lm_lin(formula = Y ~ Z, covariates = ~ x1 + x2 + x3)`

where formula is like in `difference_in_means`, just including the outcome and treatment indicator.

**Comment by lukesonnet** [comment]
I'm implementing this now. Something that's always bugged me is the inclusion of dummy variables and especially fixed effects in this estimator. I'm assuming we should go ahead and demean and interact those as well?

While I will do that and that's what I think is most sensible, it is weird that the treatment effect we are estimating is the treatment effect that is weighted by the average distribution of units across blocks for which we include fixed effects. I guess that is better than assuming the fixed effects contribute in the same way to the control and treatment potential outcomes.

**Comment by acoppock** [comment]
Your intuition is correct, we are modeling the treatment and control response surfaces completely independently.  "Fixed effect" is a red herring - it's just a covariate like anything else.

**Comment by lukesonnet** [comment]
I think it's best to write `lm_lin` and `lm_robust` as processors that create the design matrices, clusters, and weights, and both call a separate function `lm_fit` that sends this data to cpp and also computes the SEs and confidence intervals.

That way it limits the number of calls to `model.frame` and the code is more compartmentalized. The problem is you only want to center the covariates by the mean of the covariate after having listwise deleting observations that have missingness in the formula, cluster, or weighting variables.

I'm going to rewrite it this way unless there are major objections.

**Comment by acoppock** [comment]
Excellent, thanks. Go forth!

**Comment by lukesonnet** [comment]
Just want to be clear, that when I construct `X_bar`, the centered version of `X`, we would want to only use the data in the `model.frame` to get the mean.

I think this is the right behavior, I just know it will give people different answers between `lm_robust` and `lm_lin` in the following cases:

```
dat <- data.frame(Y = rnorm(100), Z = rbinom(100, 1, 0.5), X = runif(100))
dat$Z[23] <- NA

# center Xbar, but keeps 23rd observation
dat$X_bar <- dat$X - mean(dat$X)

lm_lin(Y~Z, ~X, dat)
lm_robust(Y~Z + Z*X_bar, dat)
# these will give different answers because X_bar will have been centered by a slightly different value

# also provides different answers when using subset, because lm_lin centers _after_the subset is taken
dat <- data.frame(Y = rnorm(100), Z = rbinom(100, 1, 0.5), X = runif(100))
lm_lin(Y~Z, ~X, dat, subset = Y>0)
lm_robust(Y~Z + Z*X_bar, dat, subset = Y>0)
```

I think it's unlikely that if people actually want the lin estimator, that they want to be interacting with variables that are not mean=0, even if its on a subset of the data, but I just wanted to confirm.

**Comment by lukesonnet** [comment]
Okay, this is completed, passes most tests, and finally handles missingness in the same way as lm_robust without having to call `model.frame` twice.

This is all implemented in PR #21

**Comment by graemeblair** [comment]
Sounds right! Great.
